### PR TITLE
Add the capability to determine the encoding of a resource by using BOM

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @churkin @miherlosev

### CODE REVIEW

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/node/index.js [hunk 1]
```diff
@@ -37,6 +39,19 @@ export default class NodeSandbox extends SandboxBase {
         }
     }
 
+    _parseDocumentCharset () {
+        var metaCharset = NativeMethods.querySelector.call(document, '.' + CONST.SHADOW_UI_CHARSET_CLASSNAME);
+
+        if (metaCharset) {
+            metaCharset.parentNode.removeChild(metaCharset);
```
Comment (inikulin) [inline src/client/sandbox/node/index.js]: Why we need to remove it from DOM?

File: src/client/utils/url.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import sharedUrlUtil from '../../utils/url';
+import hammerhead from '..';
```
Comment (inikulin) [inline src/client/utils/url.js]: Path should always contain at least one slash, so it's `../`

File: src/pipeline/pipeline-context.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import * as urlUtils from '../utils/url';
 import * as contentUtils from '../utils/content';
 import { XHR_REQUEST_MARKER_HEADER } from '../const';
+import ContentCharset from '../utils/content-charset';
```
Comment (inikulin) [inline src/pipeline/pipeline-context.js]: Since `ContentCharset` now is not just a bunch of utility methods it has nothing to do with the utils and should be moved to `pipeline` and renamed to `Charset` I guess.

File: src/processing/resources/manifest.js [hunk 4]
```diff
@@ -2,7 +2,7 @@ import ResourceProcessorBase from './resource-processor-base';
 import trim from '../../utils/string-trim';
 
 class ManifestProcessor extends ResourceProcessorBase {
-    processResource (manifest, ctx, charset, urlReplacer) {
+    processResource (manifest, ctx, contentCharset, urlReplacer) {
```
Comment (inikulin) [inline src/processing/resources/manifest.js]: Here and in all other places: using `charset` instead of `contentCharset` is just fine

File: src/utils/content-charset.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
+const CHARSET_RE      = /(?:^|;)\s*charset=(.+)(?:;|$)/i;
+const META_CHARSET_RE = /charset ?= ?['"]?([^ ;"']*)['"]?/i;
+// NOTE: HTTP 1.1 specifies ISO-8859-1 as a default charset
```
Comment (inikulin) [inline src/utils/content-charset.js]: Add new line before comment

File: src/utils/content-charset.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+// (see: http://www.w3.org/International/O-HTTP-charset.en.php).
+const DEFAULT_CHARSET         = 'iso-8859-1';
+const CHARSET_BOM_LIST        = [
```
Comment (inikulin) [inline src/utils/content-charset.js]: Use new lines to separate object and array properties to achieve better formatting.

File: src/utils/content-charset.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    'koi8-u', 'utf-8', 'utf-16', 'utf-16le', 'utf-16be', 'utf-32',
+    'shift-jis', 'x-euc', 'big5', 'euc-kr'
+];
+const NORMALIZED_CHARSETS_MAP = CHARSETS.reduce((charsetMap, charset) => {
+    charsetMap[getNormalizedCharsetMapKey(charset)] = charset;
+    return charsetMap;
+}, {});
+
+
+function getNormalizedCharsetMapKey (charset) {
```
Comment (inikulin) [inline src/utils/content-charset.js]: Since we have a class now, make them `static` methods

File: src/utils/content-charset.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    return charset.replace(/-/g, '').toLowerCase();
+}
+
+function normalizeCharset (charset) {
+    var key = charset && getNormalizedCharsetMapKey(charset);
+
+    return NORMALIZED_CHARSETS_MAP[key] || null;
+}
+
+function checkBOM (resBuf, bom) {
```
Comment (inikulin) [inline src/utils/content-charset.js]: checkBOM for what?

File: src/utils/content-charset.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        return true;
+    }
+
+    return false;
+}
+
+
+export default class ContentCharset {
+    constructor () {
+        this._charset  = DEFAULT_CHARSET;
```
Comment (inikulin) [inline src/utils/content-charset.js]: We don't use underline notation for the class fields

File: src/utils/content-charset.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        return false;
+    }
+
+    get () {
+        return this._charset;
+    }
+
+    isCharsetFromBOM () {
```
Comment (inikulin) [inline src/utils/content-charset.js]: `isFromBOM`

File: src/utils/content-charset.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        return this._priority === PRIORITY_LIST.BOM;
+    }
+
+    parseCharsetFromBOM (resBuf) {
```
Comment (inikulin) [inline src/utils/content-charset.js]: `parseCharsetFromBOM` - > `fromBOM`

File: src/utils/content-charset.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        for (var i = 0; i < CHARSET_BOM_LIST.length; i++) {
+            if (checkBOM(resBuf, CHARSET_BOM_LIST[i].bom))
+                return this.set(CHARSET_BOM_LIST[i].charset, PRIORITY_LIST.BOM);
+        }
+
+        return false;
+    }
+
+    parseCharsetFromContentType (contentTypeHeader) {
```
Comment (inikulin) [inline src/utils/content-charset.js]: `fromContentType`

File: src/utils/content-charset.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            var charsetMatch = contentTypeHeader && contentTypeHeader.match(CHARSET_RE);
+            var charset      = charsetMatch && charsetMatch[1];
+
+            return this.set(normalizeCharset(charset), PRIORITY_LIST.CONTENT_TYPE);
+        }
+
+        return false;
+    }
+
+    parseCharsetFromURL (charsetFromUrl) {
```
Comment (inikulin) [inline src/utils/content-charset.js]: `fromUrl`

File: src/utils/content-charset.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            return this.set(normalizeCharset(charsetFromUrl), PRIORITY_LIST.URL);
+
+        return false;
+    }
+
+    // NOTE: parsing charset from meta-tags
+    // www.whatwg.org/specs/web-apps/current-work/multipage/parsing.html#determining-the-character-encoding
+    // Each <meta> descriptor should contain values of the "http-equiv", "content" and "charset" attributes.
+    //TODO test it!!!!
+    parseCharsetFromMeta (metas) {
```
Comment (inikulin) [inline src/utils/content-charset.js]: `fromMeta`

File: test/server/content-charset-test.js [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+
+    after(function () {
+        destServer.close();
+        proxy.close();
+        requestAgent.resetKeepAliveConnections();
+    });
+
+    // Tests
+    describe('Pages', function () {
+        var charsetTestMatch  = new RegExp('<meta class="' + SHARED_CONST.SHADOW_UI_CHARSET_CLASSNAME + '" charset="([^\"]+)"', '');
```
Comment (inikulin) [inline test/server/content-charset-test.js]: What does `charsetTestMatch` mean?

File: test/server/content-charset-test.js [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+        var charsetTestFunc   = function (originUrl, expectedCharset, expectedBody, callback) {
```
Comment (inikulin) [inline test/server/content-charset-test.js]: We don't use functional expressions. `charsetTestFunc` -> `testDocumentEncoding`, `callback` -> `done`

File: test/server/content-charset-test.js [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+                var match       = decodedBody.match(charsetTestMatch);
+                var charset     = match && match[1];
+
+                expect(charset).eql(expectedCharset);
+                compareCode(body, expectedBody);
+
+                callback();
+            });
+        };
+        var emptyFunc         = function () {
```
Comment (inikulin) [inline test/server/content-charset-test.js]: It's `noop`. Search in other files how it defined

File: test/server/content-charset-test.js [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+        };
+        var getExpectedSrcStr = function (src, charset, addBOM) {
```
Comment (inikulin) [inline test/server/content-charset-test.js]: `getExpectedStr`. BTW I find it quite hard to parse this method. Please, refactor it.

File: test/server/content-charset-test.js [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+            return iconv.encode(pageProcessor.processResource(src, {}, {
+                get: function () {
+                    return charset;
+                },
+                parseCharsetFromMeta: emptyFunc
+            }, emptyFunc, {
+                scripts: [
+                    'http://' + proxyHostname + ':' + proxyPort1 + '/hammerhead.js',
```
Comment (inikulin) [inline test/server/content-charset-test.js]: Use full URL's instead of concatenation, e.g. see proxy test

File: test/server/content-charset-test.js [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+                charsetDefault,
+                getExpectedSrcStr(pageWithoutMetaSrc, charsetDefault, false),
+                done
+            );
+        });
+    });
+
+    describe('Scripts', function () {
+        var processedScript = scriptProcessor.processResource(scriptSrc);
+        var scriptTestFunc  = function (originUrl, expectedCharset, expectedBody, callback) {
```
Comment (inikulin) [inline test/server/content-charset-test.js]: `scriptTestFunc ` -> `testScriptEncoding`

File: test/server/content-charset-test.js [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+                done
+            );
+        });
+    });
+
+    describe('Other resources', function () {
+        var urlReplacer           = function (url) {
+            return urlUtil.getProxyUrl(url, proxyHostname, proxyPort1, sessionId);
+        };
+        var testOtherResourceFunc = function (resourceType, expectedBody, callback) {
```
Comment (inikulin) [inline test/server/content-charset-test.js]: `testResourceCharset`

File: test/server/content-charset-test.js [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+            var resourceUrl = 'http://127.0.0.1:2000/other-resource/' + resourceType + '/';
+            var proxyUrl    = urlUtil.getProxyUrl(resourceUrl + 'bom', proxyHostname, proxyPort1, sessionId);
+
+            request(proxyUrl, function (err, res, body) {
```
Comment (inikulin) [inline test/server/content-charset-test.js]: Promisify request and use promises instead of nested callbacks (seems like you can test url in parallel).

File: test/server/content-utils-test.js [hunk 23]
```diff
@@ -38,6 +44,37 @@ describe('Content utils', function () {
                     done();
                 });
         });
+
+        it('Should correctly determine the charset from BOM', function (done) {
```
Comment (inikulin) [inline test/server/content-utils-test.js]: It's should be moved to the `charset` tests. And shouldn't we test other parsing scenarios, e.g. meta's?

**Comment by inikulin** [comment]
\r-

File: src/client/sandbox/node/index.js (hunk 1 above)
Comment (LavrovArtem) [inline src/client/sandbox/node/index.js]: Because after parsing the meta becomes unnecessary.

Comment (inikulin) [inline src/client/sandbox/node/index.js]: Anyway, I beliveve it's a bad idea to remove it. We can do it for any our resource, but we don't, because it's useful for the testing purposes.

Comment (inikulin) [inline src/client/sandbox/node/index.js]: As In understand it's covered by shadow UI, so I don't see the reason why we can't keep it.

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/node/index.js [hunk 24]
```diff
@@ -2,6 +2,7 @@ import SandboxBase from '../base';
 import WindowSandbox from './window';
 import DocumentSandbox from './document';
 import ElementSandbox from './element';
+import NativeMethods from '../native-methods';
```
Comment (inikulin) [inline src/client/sandbox/node/index.js]: We don't use upper case notation for object anymore, it's used only for the class contructors now (see other sources). So it should be `nativeMethods`

File: src/pipeline/charset.js [hunk 25]
```diff
@@ -0,0 +1,158 @@
+const CHARSET_RE      = /(?:^|;)\s*charset=(.+)(?:;|$)/i;
+const META_CHARSET_RE = /charset ?= ?['"]?([^ ;"']*)['"]?/i;
+
+// NOTE: HTTP 1.1 specifies ISO-8859-1 as a default charset
+// (see: http://www.w3.org/International/O-HTTP-charset.en.php).
+const DEFAULT_CHARSET         = 'iso-8859-1';
+const CHARSET_BOM_LIST        = [
```
Comment (inikulin) [inline src/pipeline/charset.js]: new line here

File: src/pipeline/charset.js [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+        charset: 'utf-16le',
+        bom:     [0xFF, 0xFE]
+    },
+
+    {
+        charset: 'utf-16be',
+        bom:     [0xFE, 0xFF]
+    }
+];
+const PRIORITY_LIST           = {
```
Comment (inikulin) [inline src/pipeline/charset.js]: and here

File: src/pipeline/charset.js [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+    BOM:          3,
+    CONTENT_TYPE: 2,
+    URL:          1,
+    META:         1,
+    DEFAULT:      0
+};
+const CHARSETS                = [
```
Comment (inikulin) [inline src/pipeline/charset.js]: and here

File: src/pipeline/charset.js [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+        if (charsetFromUrl && this.priority <= PRIORITY_LIST.URL)
+            return this.set(Charset.normalizeCharset(charsetFromUrl), PRIORITY_LIST.URL);
+
+        return false;
+    }
+
+    // NOTE: parsing charset from meta-tags
+    // www.whatwg.org/specs/web-apps/current-work/multipage/parsing.html#determining-the-character-encoding
+    // Each <meta> descriptor should contain values of the "http-equiv", "content" and "charset" attributes.
+    //TODO test it!!!!
```
Comment (inikulin) [inline src/pipeline/charset.js]: We need test for this as `TODO` suggests

File: src/processing/resources/index.js [hunk 29]
```diff
@@ -37,8 +38,12 @@ export async function process (ctx, customCharset) {
             var urlReplacer = getResourceUrlReplacer(ctx);
             var processed   = processors[i].processResource(decoded, ctx, charset, urlReplacer);
 
-            if (processed === null)
+            if (processed === null) {
+                if (processors[i] === pageProcessor)
```
Comment (inikulin) [inline src/processing/resources/index.js]: This is potentially infinite recursion. We need better mechanism and more clear mechanism to tell `process` reprocess resource with the modified `charset`

File: src/processing/resources/page.js [hunk 30]
```diff
@@ -118,40 +106,38 @@ class PageProcessor extends ResourceProcessorBase {
 
         var $ = whacko.load(html);
 
-        actualCharset = actualCharset || defaultCharset;
-
-        var rightCharset = this._getRightCharset($, defaultCharset, actualCharset);
+        if (charset.fromMeta(this._getPageMetas($))) {
+            //restart injecting
+            this.processResource(html, ctx, charset, urlReplacer, processingOpts);
```
Comment (inikulin) [inline src/processing/resources/page.js]: Why we call `this.processResource` here and instantly return `null` which tells caller to restart whole processing?

File: test/server/charset-test.js [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+
+describe('Content charset', function () {
+    var destServer = null;
+    var proxy      = null;
+    var session    = null;
+
+    var proxyHostname = '127.0.0.1';
+    var proxyPort1    = 1836;
+    var proxyPort2    = 1837;
+    var sessionId     = 'sessionId';
```
Comment (inikulin) [inline test/server/charset-test.js]: Why we need this vars block? Just pass values directly where necessary

File: test/server/charset-test.js [hunk 32]
```diff
@@ -0,0 +1,329 @@
...
+    var pageWithoutMetaSrc = fs.readFileSync('test/server/data/content-charset/page-without-meta.htm').toString();
+    var scriptSrc          = fs.readFileSync('test/server/data/content-charset/script.js').toString();
+    var manifestSrc        = fs.readFileSync('test/server/data/content-charset/manifest').toString();
+    var stylesheetSrc      = fs.readFileSync('test/server/data/content-charset/style.css').toString();
+
+    var charsetBOM         = 'utf-16be';
+    var charsetContentType = 'utf-8';
+    var charsetMeta        = 'windows-1251';
+    var charsetDefault     = 'iso-8859-1';
+    var charsetUrl         = 'utf-16le';
```
Comment (inikulin) [inline test/server/charset-test.js]: Same here, just pass them directly, moreover their naming is not semantically correct.

File: test/server/charset-test.js [hunk 33]
```diff
@@ -0,0 +1,329 @@
...
+
+    after(function () {
+        destServer.close();
+        proxy.close();
+        requestAgent.resetKeepAliveConnections();
+    });
+
+    // Tests
+    describe('Pages', function () {
+        var testDocumentEncoding = function (originUrl, expectedCharset, expectedBody, done) {
```
Comment (inikulin) [inline test/server/charset-test.js]: Once again: **we don't use functional expressions**

File: test/server/charset-test.js [hunk 34]
```diff
@@ -0,0 +1,329 @@
...
+        };
+        var getExpectedSrcStr    = function (src, charsetStr, addBOM) {
+            var charset           = {
+                get: function () {
+                    return charsetStr;
+                },
+
+                fromMeta: noop
+            };
+            var proxyResources    = {
```
Comment (inikulin) [inline test/server/charset-test.js]: Add new line here, to correct formatting. Always separate object and array vars with the new line

File: test/server/charset-test.js [hunk 35]
```diff
@@ -0,0 +1,329 @@
...
+
+            request(options, function (err, res, body) {
+                compareCode(body, expectedBody);
+
+                done();
+            });
+        };
+        var noop                 = function () {
+        };
+        var getExpectedSrcStr    = function (src, charsetStr, addBOM) {
```
Comment (inikulin) [inline test/server/charset-test.js]: `getExpectedStr`

File: test/server/charset-test.js (hunk 33 above)
Comment (inikulin) [inline test/server/charset-test.js]: Let's call it `testDocumentCharset`, my bad

File: test/server/charset-test.js [hunk 36]
```diff
@@ -0,0 +1,329 @@
...
+                charsetDefault,
+                getExpectedSrcStr(pageWithoutMetaSrc, charsetDefault, false),
+                done
+            );
+        });
+    });
+
+    describe('Scripts', function () {
+        var processedScript = scriptProcessor.processResource(scriptSrc);
+        var scriptTestFunc  = function (originUrl, expectedCharset, expectedBody, done) {
```
Comment (inikulin) [inline test/server/charset-test.js]: `testScriptCharset`

File: test/server/charset-test.js [hunk 37]
```diff
@@ -0,0 +1,329 @@
...
+    });
+
+    it('Should correctly determine the charset from BOM', function (done) {
+        var testBOM = function (bomCharset, contentTypeHeader) {
+            var inputData = iconv.encode(pageWithoutMetaSrc, bomCharset, { addBOM: true });
+            var charset   = new Charset();
+
+            charset.fromContentType(contentTypeHeader);
+
+            return contentUtils.decodeContent(inputData, '', charset).then(function (decoded) {
```
Comment (inikulin) [inline test/server/charset-test.js]: In all promises - place `then` on the new line

File: test/server/charset-test.js [hunk 38]
```diff
@@ -0,0 +1,329 @@
...
+                expect(decoded).eql(pageWithoutMetaSrc);
+                expect(charset.get()).eql(bomCharset);
+                expect(charset.isFromBOM()).to.be.true;
+
+                return contentUtils.encodeContent(pageWithoutMetaSrc, '', charset).then(function (encoded) {
```
Comment (inikulin) [inline test/server/charset-test.js]: It can be refactored to avoid additional nesting level:
```js
  return contentUtils.decodeContent(inputData, '', charset)
     .then(function(decoded) {
         expect(decoded).eql(pageWithoutMetaSrc);
         expect(charset.get()).eql(bomCharset);
         expect(charset.isFromBOM()).to.be.true;

         return contentUtils.encodeContent(pageWithoutMetaSrc, '', charset);
     })
     .then(function(encoded) {
         expect(encoded).to.deep.equal(inputData);
     });
```

File: test/server/charset-test.js [hunk 39]
```diff
@@ -0,0 +1,329 @@
...
+                    testResourceCharset(processedStylesheet, charsetDefault, getProxyUrl(resourceUrl + 'default'))
+                ])
+                .then(function () {
+                    done();
+                })
+                .catch(done);
+        });
+    });
+
+    it('Should correctly determine the charset from BOM', function (done) {
```
Comment (inikulin) [inline test/server/charset-test.js]: We still missing test for parsing from metas and URL (see above)

**Comment by inikulin** [comment]
\r-

File: test/server/charset-test.js (hunk 39 above)
Comment (LavrovArtem) [inline test/server/charset-test.js]: parsing from metas
```javascript
it('Should correctly determine the charset from meta', function () {
```
parsing from url
```javascript
describe('Scripts', function () {
```

**Comment by LavrovArtem** [comment]
FPR

**Comment by LavrovArtem** [comment]
ping @churkin @miherlosev

**Comment by miherlosev** [comment]
lgtm

File: src/pipeline/charset.js [hunk 40]
```diff
@@ -0,0 +1,160 @@
...
+        if (resBuf.length >= bom.length) {
+            for (var i = 0; i < bom.length; i++) {
+                if (resBuf[i] !== bom[i])
+                    return false;
+            }
+
+            return true;
+        }
+
+        return false;
```
Comment (inikulin) [inline src/pipeline/charset.js]: It's better be rewritten as:
```js 
static checkSourceBOM(resBuf, bom) {
     if(resBuf.length < bom.length)
       return false;
   
     for (var i = 0; i < bom.length; i++) {
         if (resBuf[i] !== bom[i])
             return false;
     }

     return true;
 }
```

`checkSourceBOM` is still quite unclear name. It should be `startsWithBOM` or smthg like that.

File: src/pipeline/charset.js [hunk 41]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    isFromBOM () {
+        return this.priority === PRIORITY_LIST.BOM;
+    }
+
+    fromBOM (resBuf) {
+        for (var i = 0; i < CHARSET_BOM_LIST.length; i++) {
+            if (Charset.checkSourceBOM(resBuf, CHARSET_BOM_LIST[i].bom))
+                return this.set(CHARSET_BOM_LIST[i].charset, PRIORITY_LIST.BOM);
```
Comment (inikulin) [inline src/pipeline/charset.js]: It's inconsistent. It should return the operations status (`true | false`) or shouldn't return anything.

File: src/processing/resources/index.js [hunk 42]
```diff
@@ -37,8 +37,8 @@ export async function process (ctx, customCharset) {
             var urlReplacer = getResourceUrlReplacer(ctx);
             var processed   = processors[i].processResource(decoded, ctx, charset, urlReplacer);
 
-            if (processed === null)
-                return null;
+            if (processed && processed.err)
+                return await process(ctx, processed.rightCharset);
```
Comment (inikulin) [inline src/processing/resources/index.js]: Why we need to pass charset and save it to `rightCharset`. AFAIK it's same object what is in the `ctx.contentInfo.charset`, so `process` don't need additional argument. You can use return `restartProcessingSymbol` (https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Symbol) to notify `process` what processing should be restarted.

Comment (inikulin) [inline src/processing/resources/index.js]: \cc @churkin @AlexanderMoskovkin Maybe you will have better idea for this mechanism

**Comment by inikulin** [comment]
\r-

File: src/pipeline/charset.js (hunk 40 above)
Comment (LavrovArtem) [inline src/pipeline/charset.js]: ```sourceStartsWithBOM``` ?

Comment (inikulin) [inline src/pipeline/charset.js]: What is `source` in this context?

File: src/pipeline/charset.js (hunk 41 above)
Comment (LavrovArtem) [inline src/pipeline/charset.js]: the ```set``` function returns the operations status (```true | false```)

File: src/pipeline/charset.js (hunk 40 above)
Comment (LavrovArtem) [inline src/pipeline/charset.js]: may be ```bufferStartsWithBOM```?

**Comment by LavrovArtem** [comment]
FPR

File: src/processing/resources/index.js [hunk 43]
```diff
@@ -37,8 +38,8 @@ export async function process (ctx, customCharset) {
             var urlReplacer = getResourceUrlReplacer(ctx);
             var processed   = processors[i].processResource(decoded, ctx, charset, urlReplacer);
 
-            if (processed === null)
-                return null;
+            if (pageProcessor.isRestartProcessingSymbol(processed))
```
Comment (inikulin) [inline src/processing/resources/index.js]: Why not just make `RESTART_PROCESSING` field of the `pageProcessor` and just compare it  directly here, IMHO it will be much clear

File: src/processing/resources/page.js [hunk 44]
```diff
@@ -14,6 +13,7 @@ const BODY_CREATED_EVENT_SCRIPT =
                var script = document.currentScript || document.scripts[document.scripts.length - 1];
                script.parentNode.removeChild(script);
           </script>`;
+const RESTART_PROCESSING_SYMBOL = Symbol();
```
Comment (inikulin) [inline src/processing/resources/page.js]: Just call it `restartProcessing`. Use `new` keyword for construction BTW

File: src/processing/resources/page.js [hunk 45]
```diff
@@ -14,6 +13,7 @@ const BODY_CREATED_EVENT_SCRIPT =
                var script = document.currentScript || document.scripts[document.scripts.length - 1];
                script.parentNode.removeChild(script);
           </script>`;
```
Comment (inikulin) [inline src/processing/resources/page.js]: Just found out what that: Since it's multiline string it will include indentation, so, then injected it will have some really broken indentation. To reduce amount of bytes passed to client and because of aesthetical reason use tag for templates from [dedent](https://www.npmjs.com/package/dedent) package.

Comment (inikulin) [inline src/processing/resources/page.js]: Also use template on https://github.com/LavrovArtem/testcafe-hammerhead/blob/master/src/processing/resources/page.js#L86 and use template concatenation, to avoid new line in resulting string in https://github.com/LavrovArtem/testcafe-hammerhead/blob/master/src/processing/resources/page.js#L86 and https://github.com/LavrovArtem/testcafe-hammerhead/blob/master/src/processing/resources/page.js#L62

**Comment by inikulin** [comment]
Just some minor issues left.
\r-

File: src/processing/resources/page.js (hunk 44 above)
Comment (LavrovArtem) [inline src/processing/resources/page.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol
```javascript
var sym = new Symbol(); // TypeError
```

File: src/processing/resources/page.js (hunk 45 above)
Comment (LavrovArtem) [inline src/processing/resources/page.js]: I will do for this new "Refactoring" pull request.

**Comment by LavrovArtem** [comment]
FPR

File: src/processing/resources/page.js [hunk 46]
```diff
@@ -22,6 +21,8 @@ class PageProcessor extends ResourceProcessorBase {
         this.domProcessor = new DomProcessor(new DomAdapter());
     }
 
+    restartProcessing = Symbol();
```
Comment (inikulin) [inline src/processing/resources/page.js]: We don't use class properties anymore because more likely they will not be in the spec (see #29), so move it to ctor, use const notation and `new` keyword:
```js
this.RESTART_PROCESSING = new Symbol();
```

**Comment by inikulin** [comment]
\r-

File: src/processing/resources/page.js (hunk 46 above)
Comment (LavrovArtem) [inline src/processing/resources/page.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol
```javascript
var sym = new Symbol(); // TypeError
```

Comment (inikulin) [inline src/processing/resources/page.js]: ok

**Comment by LavrovArtem** [comment]
FPR

**Comment by inikulin** [comment]
lgtm
@churkin Seems like your PR with constants will require rebase on top of this and we have a new constants now (e.g. for charset meta)

**Comment by LavrovArtem** [comment]
@churkin FPR

**Comment by churkin** [comment]
lgtm
